CONTRIBUTING.md

```diff
@@ -0,0 +1,123 @@
+  `main`)
+
+
+## GitHub Issues & Labels
+
+Create new Github Issues using the templates wherever possible.
+
+## PR Review & Committing code
+
+Multiple commits or PRs can be created for an Issue. e.g. each implementation step might get its own PR.
+
+Code review is encouraged as a powerful tool for learning.  Benefits include
+- Spread knowledge of the code base throughout the team: reviewing code is a remarkably effective way to learn the codebase.
+- Expose everyone to different approaches.
+- Ensure code is readable (and therefore maintainable).
+- Yield better software (but ultimately the responsibility
+  for bug-free code is on the code author, not the reviewer).
+
+Code review is not limited to approval/rejection of PRs. Also consider involving a collaborator
+earlier in the process, before the code is finished.  Ask them for a narrower review—e.g., a
+design review or to focus on a specific part of the code change.
+
+As a reviewer:
+- be kind & helpful, but do not be "nice" for the sake of avoiding conflict.
+- "I find this very hard to follow" is valid feedback, even if the code's behavior is technically correct.
+- Ability for code review to find defects diminishes with longer PRs: Feel free to reject any
+  review that adds more than 400 lines of new code. (no upper limit on deletions!)
+
+Merging branches and PRs to `main`:
+- The branch author should be the one to merge.
+- Merge PRs via 'Merge Commit' option in GitHub.
+- Delete a branch when you are done with it.
+
+
+# Tests
+
+## Testing Strategy
+
+Keep the test suite
+* **complete**: every feature needs a test.
+* **focused**: test behavior, not implementation.
+* **stable**: tests should be deterministic and fully repeatable.
+* **fast**: no sleeps, no time-based tests.
+
+## CircleCI
+
+All tests in the repo should be executed by CircleCI (see `.circleci/config.yml`)
+automatically upon push to any branch.  (This can be disabled on a per-commit basis
+by appending a **`[ci skip]`** line to the commit message.)
+
+Do not merge branches if CircleCI is not green.
+
+
+# Coding & Documentation Style
+
+## Prettier & ESLint for JavaScript, JSON, & SCSS
+
+All developers are expected to use [prettier](https://prettier.io/) to format their JavaScript or TypeScript scripts in this repository.
+
+> To apply Prettier autoformatting, you can configure your text editor to format on save [according
+> to these docs](https://prettier.io/docs/en/editors.html)
+> - For VS Code, set `editor.formatOnSave` to true in your editor config.
+
+
+ESLint helps identify problematic patterns found in JavaScript code.
+
+> Follow [ESLint Getting Started](https://eslint.org/docs/user-guide/getting-started).
+> To configure ESLint in VS Code, create a `.vscode/settings.json` in the project's root and add
+> ```
+>    "eslint.nodePath": "/home/\${user}/dev/nosh-by-gosh/client/node_modules",
+>    "eslint.workingDirectories": ["client"]
+> ```
+
+
+## Ruff for Python
```

Comment:
Windmill's in-app code editor also does its own linting and ruff auto-formatting.  I have not yet read up on much more than that.

But it occurs to me that whatever auto-formatting Windmill's editor does will need to be matched exactly with local-dev formatting.  For example I am pretty sure that Windmill sets its ruff max line-length to 88 (maybe this is the default, again I haven't looked).

Any ideas how to easily make these sync up?  Some ideas from my end:
- Are windmill's default lint/formatter settings in their version control? We could copy those files.
- Can we set some file in our repo that overrides Windmill's settings?
- Does the `wmill` CLI have anything that formats local code on the windmill server, according to their own rules?


It would be good for one of us to read up on whatever docs or old GH issues Windmill has around this.

Reply:
[In the codebase](https://github.com/search?q=repo%3Awindmill-labs%2Fwindmill%20ruff&type=code) it just looks like they run `pip3 install ruff-lsp` in the Dockerfile and provide it as a service through a WebSocket endpoint. 

I asked Windmill's AI service on their Discord:

----

Answer
Based on the available information sources, while Windmill uses Ruff, Black, and Pyright for Python (1)  (2) , the specific linting and formatting settings for these tools are not documented in the provided sources. 

From what we can see in one debug log, Ruff is run with these flags (3) :
```
['--force-exclude', '--no-cache', '--no-fix', '--quiet', '--output-format', 'json', '-', '--stdin-filename']
```

However, I cannot make any claims about other configuration settings as they are not explicitly documented in the information sources. If you need to know about specific linting rules or formatting configurations, I'd suggest clicking the "Ask for help" button to have a team member provide more details.

Sources
[[1, 2] Code assistants](https://www.windmill.dev/docs/code_editor/assistants)
[[3] can not run any python script](https://github.com/windmill-labs/windmill/discussions/3320)

Reply:
Wow, thank you so much for doing this digging.

Upon further fiddling with the Windmill in-browser code editor ... it auto-formats Javascript but not Python!  It does provide lint and typehints, but no auto-formatting.

And as you found it doesn't seem that windmill-lsp overrides any defaults.

So -- I believe that what CONTRIBUTING.md says is correct and will not conflict with anything!